### PR TITLE
Fix preserveSourceNewlines sibling node comparison (fixes extra newlines in organize imports)

### LINKED ISSUE #41417: Typescript "Organize Imports" action inserts loads of blank lines in import statements
*TS Template added by @mjbvz*

**TypeScript Version**: 4.2.0-dev.20201103

**Search Terms**
- organize imports
- alias

---

When running TypeScript's "Organize Imports" action on this file from the LSP client:

https://github.com/microsoft/vscode-languageserver-node/blob/05fa3b55538acaa87647111d528d39ddbd86caa2/client/src/common/client.ts

It makes a mess of the file, inserting lots of blank lines in between each class:

Nov-05-2020 14-23-21

I've disabled all other extensions but it still occurs. It shows TS 4.0.3 in the status bar if that's important.

### PR DESCRIPTION
**Pull request by andrewbranch**
This is @armanio123’s #41927 with an extra two commits on top.

Fixes #41417.

### CODE REVIEW

File: tests/baselines/reference/parserErrorRecovery_ObjectLiteral2.js [hunk 1]
```diff
@@ -3,4 +3,5 @@ var v = { a
 return;
 
 //// [parserErrorRecovery_ObjectLiteral2.js]
-var v = { a: a, "return":  };
+var v = { a: a,
+    "return":  };
```
Comment (andrewbranch) [inline tests/baselines/reference/parserErrorRecovery_ObjectLiteral2.js:7]: These JS baseline changes are reverting back to their original state pre-https://github.com/microsoft/TypeScript/pull/36688. That PR was never really meant to change JS emit, but the couple changes that did occur seemed ok so we went with it. Since `siblingNodePositionsAreComparable` might do a little more work than what’s currently happening in master, I reverted the codepath that non-codefix/refactor emit takes back to the way it originally was.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -7119,6 +7119,39 @@ namespace ts {
...
+        const getContainingChild = (child: Node) => {
+            if (child.pos <= position && (position < child.end || (position === child.end && (child.kind === SyntaxKind.EndOfFileToken)))) {
+                return child;
+            }
+            else if (!next && child.pos > position) {
+                next = child;
+            }
+        };
+        while (true) {
+            const child = isSourceFileJS(sourceFile) && hasJSDocNodes(current) && forEach(current.jsDoc, getContainingChild) || forEachChild(current, getContainingChild);
```
Comment (rbuckton) [inline src/compiler/utilities.ts:7136]: Is this doing a recursive walk of a Node's entire subtree? That sounds expensive for something could possibly call between every list element.

**Comment by rbuckton** [comment]
@typescript-bot perf test

File: src/compiler/utilities.ts (hunk 2 above)
Comment (andrewbranch) [reply src/compiler/utilities.ts:7136]: This only ever happens when printing nodes from `TextChanges`.

Comment (andrewbranch) [reply src/compiler/utilities.ts]: Nonetheless, I reworked to make it start at the nearest possible ancestor instead of the source file, and skipped it altogether in cases where neither list element went through a transformation (which is the only way they could be reordered or reparented, I think).

**Comment by rbuckton** [comment]
Looks like the bot trigger for perf tests didn't pick up my command due to a typo in the regexp.

@typescript-bot perf test this

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -4520,16 +4520,18 @@ namespace ts {
                 return false;
             }
 
+            // If nodes haven't been transformed, they should always be real siblings
+            if (!previousNode.original && !nextNode.original) {
```
Comment (rbuckton) [inline src/compiler/emitter.ts:4524]: `original` is only set when using a `factory.updateX` method or when calling `setOriginalNode`, so this isn't exactly the best test. As written, this only tests whether an `original` pointer has not been set, which is also true for any `factory.createX` method in any transform.

If you want to _only_ focus on parse tree nodes (i.e., nodes that were created by the parser), use `isParseTreeNode` or check for `!(node.flags & NodeFlags.Synthesized)`.

We have a bunch of conflicting terms around `Node` objects that I've been meaning to clarify. Functions like `nodeIsSynthesized` aren't *quite* accurately named, since they are checking for "non-existent positions" (i.e., `pos` or `end` is `-1`). The function really should have been something like `textRangeIsMissing` or something.

Terminology wise:
- A "Parse Tree Node" is a `Node` produced by the parser (`(node.flags & NodeFlags.Synthesized) === 0`).
  - A "Parse Tree Node" will *always* have a valid (i.e. `>= 0`) `pos` and `end` (though they may be the same, indicating a missing node).
  - A "Parse Tree Node" will have a `parent` pointer if:
    - The node is not a `SourceFile` (they don't have parents)
    - The containing `SourceFile` was either parsed with `true` for `setParentNodes`, or was bound.
- A "Synthetic Node" is a `Node` produced outside of the parser (`(node.flags & NodeFlags.Synthesized) !== 0`).
  - A "Synthetic Node" *may* have valid `pos` and `end` values for various reasons:
    - A "Synthetic Node" created from an `update` method will copy the `pos` and `end` of its original node.
    - A "Synthetic Node" may have its `pos` and `end` copied from a source node via `setTextRange`.
      - This is often the case for copies of an `Identifier` or literal that came from the parse tree as a means of preserving the quote style or escape characters used in the original source. The copies are often used when the node needs to be used twice with different source map or comment handling. 
    - A "Synthetic Node" can have a mix of valid (`>=0`) or invalid (`-1`) values for `pos` and `end`.
- An "Original Node" is a `Node` that was the source of a transformation (often through calling an `update` method), but an "Original Node" could be *either* a "Parse Tree Node" or a "Synthetic Node".
  - Some "Parse Tree Nodes" *will* have `original` pointers due to some cases in `Parser` where we reparse a node (such as with Top-Level Await). There will likely be more of these cases as at some point I'd like to introduce `ObjectAssignmentPattern` and `ArrayAssignmentPattern` as a reparse of `ObjectLiteralExpression` and `ArrayLiteralExpression` in assignment patterns (our current approach for handling those in destructuring is overly complex).

tldr: If the goal here is to only perform this check on adjacent "Parse Tree Nodes" that belong to the same parent, then you probably want to use `isParseTreeNode(previousNode) && isParseTreeNode(nextNode) && previousNode.parent === nextNode.parent`.

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -7119,10 +7119,11 @@ namespace ts {
         }
     }
 
-    export function nodeIsFirstNodeAtOrAfterPosition(sourceFile: SourceFile, node: Node, position: number): boolean {
+    export function nodeIsFirstNodeAtOrAfterPosition(node: Node, position: number): boolean {
```
Comment (rbuckton) [inline src/compiler/utilities.ts:7122]: Given my comment for `siblingNodePositionsAreComparable` above, if this function is limited to "adjacent parse tree nodes belonging to the same parent", you can determine if they are truly adjacent (i.e., their order hasn't changed) in two much less expensive ways:

1. Examine the `end` of `previousNode` and the `pos` of `nextNode` and use the context of the parent to determine what, if any, separator token may be present. You can then use `getSourceFileOfNode` (since these nodes are parse tree nodes), `skipTrivia`, and the size of the separator token (if present) to check adjacency.
2. Alternatively, (and possibly even more cheaply) you can use the context of the `parent` node to find which `NodeArray` each node belongs two and use `indexOf` to test for positions. While some of our `Nodes` have multiple `NodeArray` properties, they don't overlap (i.e., for a function-like, `Expression` nodes are probably `arguments`, while `TypeNode` nodes are probably `typeArguments`, etc.).

File: src/compiler/utilitiesPublic.ts [hunk 5]
```diff
@@ -1046,7 +1046,7 @@ namespace ts {
     // All node tests in the following list should *not* reference parent pointers so that
     // they may be used with transformations.
     /* @internal */
-    export function isNode(node: Node) {
+    export function isNode(node: any): node is Node {
```
Comment (rbuckton) [inline src/compiler/utilitiesPublic.ts:1049]: That's not what this function does. Its a misclassification, unfortunately. This function tests whether a `Node` is "a `Node` that is not a `Token`" (i.e., is has `kind >= SyntaxKind.QualifiedName`). It should probably have been named `isNonTokenNode`, tbh. If you want to check whether something is a `Node`, that is a little trickier. You'd have to check that: 
- the argument is an `object` that is not `null`.
- the argument has `kind`, `pos`, and `end` fields (at a minimum).

Alternatively you'd need to use the *current* value of `objectAllocator` and check whether the value is an `instanceof` an of `objectAllocator.getNodeConstructor()`, `.getTokenConstructor()`, `.getIdentifierConstructor()`, `.getPrivateIdentifierConstructor()`, or `.getSourceFileConstructor()`. You'd have to use the *current* value since the language service overwrites `objectAllocator` with the more complex version.

Either way, I'd recommend a different function for that.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: src/compiler/utilities.ts (hunk 4 above)
Comment (andrewbranch) [reply src/compiler/utilities.ts:7122]: I had thought of something like both of those before. I think option (2) is probably best; it just requires writing a tediously large switch statement on the parent kind. But maybe not too tedious and not too large. For option (1), I don’t think size is sufficient to tell whether something is a separator or not:

```ts
interface I {
  a
  b
  c
}
```

Are `a` and `c` adjacent? There’s only one character of non-trivia between them, but I don’t know whether it’s actually a comma/semicolon or another element without actually looking at the text between them, which seemed more tedious than the former approach. I also do think `skipTrivia` will probably be more expensive than option (1).

File: src/compiler/utilitiesPublic.ts (hunk 5 above)
Comment (andrewbranch) [reply src/compiler/utilitiesPublic.ts:1049]: Oops, I guess I didn’t check what `isNodeKind` actually does, and didn’t end up needing this change after all anyway. But do note, this function is literally never used.

File: tests/baselines/reference/parserErrorRecovery_ObjectLiteral2.js (hunk 1 above)
Comment (andrewbranch) [reply tests/baselines/reference/parserErrorRecovery_ObjectLiteral2.js:7]: `siblingNodePositionsAreComparable` is now quite cheap, but I think it still makes sense to keep it out of JS emit for now.

**Review by RyanCavanaugh — APPROVED** [review]
